crypto/src/node.rs

```diff
@@ -81,6 +83,18 @@ impl<T> NodeMap<T> {
         NodeMap(v)
     }
 
+    pub fn with_hashmap(len: NodeCount, hashmap: HashMap<NodeIndex, T>) -> Self
```

Comment:
Should probably be `from_hashmap` (unless clippy complains :/), afaik `with` implies using the thing for some behind-the-scenes stuff, like allocators for `Vec`s etc.

Reply:
possibly the best way would be to implement `Extend` and then construct the `NodeMap` like that: `NodeMap::with_size(len).extend(hashmap)` - what do you think?

Reply:
That does look quite cool, but I would use a different name (maybe `fill`?) – `extend` doesn't make much sense for fixed size objects.

Reply:
but isn't the fixed "size" just an implementation detail? for me, `NodeMap` pretends to be a `HashMap` with additional constraints (other than the type) on keys?

have you seen any similar structs in other crates? e.g. `ArrayVec` and `FcHashMap` require the capacity to be known at compile time, so it's not quite it

then again, we probably _do_ know the number of nodes at compile time, don't we?

Reply:
We don't know it at compile time actually, only at object creation time. But after the object is created, its size cannot change, so `extend` doesn't make sense. In particular, this method should drop any thinks in the `NodeMap` that have indices associated with nodes outside of the acceptable range, which is not expected behaviour for `extend`.

This would be a perfect usecase for dependent types, but unfortunately we are not writing in Agda. :'<

Reply:
I meant that we probably know how many committee members we have at compile time :P

And again, here "size" is somehow misleading, as it should be called "capacity". Both structs I've mentioned above have fixed capacity, but also implement `len` and `extend`. Imho `extend` makes perfect sense if `len` can change.

Reply:
if `from_hashmap` is not possible then any of these is fine...

Reply:
`len` cannot really change after a `NodeMap` object is created, so `extend` does not sound good to me.

Reply:
ok, that's bad, because I've just added `len` that _can_ and _will_ change - please look at the implementation and give some feedback before I merge this PR

Reply:
The whole abstraction here is "we are looking at something indexed by a set which we don't know at compile time, but at runtime is fixed". This doesn't quite map to what any of the structs you mentioned are trying to do, and I would argue that using language standardized for them ("capacity", "length", "extend" etc.) is likely to be confusing.

And no, we don't know the committee size at compile time, even now we could change it without recompiling anything and it definitely should stay that way...

Reply:
also, `from_hashmap` is possible, but the point is that I'd personally expect this method to have only a single argument, whereas here we'd also need to provide the capacity

Reply:
> The whole abstraction here is "we are looking at something indexed by a set which we don't know at compile time, but at runtime is fixed". This doesn't quite map to what any of the structs you mentioned are trying to do

I totally agree - but then could you provide any better examples?

Also, I think that for a struct implementing `iter` any other implementation of `len` will be much more confusing...

Reply:
(by "other" I mean "not equivalent to `.iter().count()`")

Reply:
If `from_hashmap` had only one argument it should be implemented as `From<HashMap>`, so I would say multiple args are exactly the usecase we want for a function with such a name.

`len` being equivalent to `iter().count` sounds like a reasonable policy to me.

Reply:
I don't think you will find examples among other structs. `NodeMap` is special -- it represents the mathematical concept of a partial function [N] -> T. When creating a `NodeMap` you naturally need to specify `N` (a number) and `T` (a type). 

What I don't like is that we added implementations of `iter()`, `len()`, `size()` to it, and now these things are pairwise contradictory and not consistent with the usual intuition behind these names. I would get rid of these implementations and just use custom names like `ground_set_size()` or `iter_pairs` etc.

Reply:
On top of that, I like `from_hashmap` with two args.

Reply:
The most obvious replacement for `size` is `node_count`, I think we either have a trait for this or at least are using it very consistently around the codebase. For the rest I don't have strong opinions, but making the names more obviously unrelated to the common usage sounds good.

Reply:
I'll go with `from_hashmap`, but I need to have `len` and `is_empty` (at least clippy was complaining about the second one) - @DamianStraszak how should I name those new methods then?

Reply:
The version I'm seeing has both `len` and `is_empty`. <_<

Reply:
Wait, nvm, this is `NodeSubset`. ^^"